File: UK_news_scraper/profiles.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timezone
from enum import Enum
import hashlib
import json
import os
from pathlib import Path
import re
import shutil
import sys
from typing import Any

from .config import AGENCIES, TOPIC_RULES


PROFILE_SCHEMA_VERSION = 1
DEFAULT_PROFILE_ID = "uk-tech-law"
PARLIAMENT_SOURCE_ID = "UK Parliament"
SOURCE_ID_MIGRATIONS = {
    "DSIT": ("BIST", "DCMS", "Cabinet Office"),
    "DBT": ("BIST",),
}


class KeywordStrength(str, Enum):
    CORE = "core"
    GENERAL = "general"
    SUPPORTING = "supporting"

    @property
    def label(self) -> str:
        return {
            self.CORE: "核心",
            self.GENERAL: "一般",
            self.SUPPORTING: "輔助",
        }[self]


@dataclass(frozen=True)
class KeywordDefinition:
    phrase: str
    strength: KeywordStrength = KeywordStrength.GENERAL


@dataclass(frozen=True)
class ProfileTopic:
    name: str
    keywords: tuple[KeywordDefinition, ...]


@dataclass(frozen=True)
class FilterProfile:
    profile_id: str
    name: str
    description: str
    version: int
    selected_sources: tuple[str, ...]
    topics: tuple[ProfileTopic, ...]
    minimum_score: int = 3

    @property
    def is_default(self) -> bool:
        return self.profile_id == DEFAULT_PROFILE_ID


@dataclass(frozen=True)
class ProfileLoadReport:
    profiles: dict[str, FilterProfile]
    warning: str = ""
    recovery_path: Path | None = None


_SUPPORTING_KEYWORDS = {
    "copyright",
    "evaluation",
    "innovation",
    "platform",
    "resilience",
    "supply chain",
    "technology",
}


def default_profile() -> FilterProfile:
    topics = tuple(
        ProfileTopic(
            name=rule.name,
            keywords=tuple(
                KeywordDefinition(keyword, _default_strength(keyword))
                for keyword in rule.keywords
            ),
        )
        for rule in TOPIC_RULES
    )
    sources = tuple(agency.short_name for agency in AGENCIES) + (PARLIAMENT_SOURCE_ID,)
    return FilterProfile(
        profile_id=DEFAULT_PROFILE_ID,
        name="UK 科技法制",
        description="英國科技、AI、資料治理、平台、資安、半導體與量子政策觀測",
        version=PROFILE_SCHEMA_VERSION,
        selected_sources=sources,
        topics=topics,
        minimum_score=3,
    )


def validate_profile(profile: FilterProfile) -> FilterProfile:
    if profile.version != PROFILE_SCHEMA_VERSION:
        raise ValueError(f"不支援的設定檔版本：{profile.version}")
    if not re.fullmatch(r"[a-z0-9][a-z0-9-]{1,48}", profile.profile_id):
        raise ValueError("設定檔 ID 只能使用小寫英數與連字號，長度 2 至 49 字元")
    if not profile.name.strip():
        raise ValueError("設定檔名稱不可空白")
    if profile.minimum_score < 1:
        raise ValueError("最低分數必須大於等於 1")
    if not profile.selected_sources:
        raise ValueError("至少選擇一個資料來源")
    available_sources = {agency.short_name for agency in AGENCIES} | {PARLIAMENT_SOURCE_ID}
    unknown_sources = set(profile.selected_sources) - available_sources
    if unknown_sources:
        raise ValueError(f"設定檔包含未知來源：{sorted(unknown_sources)}")
    if not profile.topics:
        raise ValueError("至少建立一個主題")

    seen_phrases: set[str] = set()
    for topic in profile.topics:
        if not topic.name.strip():
            raise ValueError("主題名稱不可空白")
        if not topic.keywords:
            raise ValueError(f"主題「{topic.name}」至少需要一個關鍵詞")
        for keyword in topic.keywords:
            phrase = keyword.phrase.strip()
            if not phrase:
                raise ValueError(f"主題「{topic.name}」包含空白關鍵詞")
            normalized = phrase.casefold()
            if normalized in seen_phrases:
                raise ValueError(f"關鍵詞重複：{phrase}")
            seen_phrases.add(normalized)
    return profile


def profile_hash(profile: FilterProfile) -> str:
    payload = json.dumps(
        profile_to_dict(profile),
        ensure_ascii=False,
        sort_keys=True,
        separators=(",", ":"),
    ).encode("utf-8")
    return hashlib.sha256(payload).hexdigest()


def profile_to_dict(profile: FilterProfile) -> dict[str, Any]:
    return {
        "profile_id": profile.profile_id,
        "name": profile.name,
        "description": profile.description,
        "version": profile.version,
        "selected_sources": list(profile.selected_sources),
        "topics": [
            {
                "name": topic.name,
                "keywords": [
                    {
                        "phrase": keyword.phrase,
                        "strength": keyword.strength.value,
                    }
                    for keyword in topic.keywords
                ],
            }
            for topic in profile.topics
        ],
        "minimum_score": profile.minimum_score,
    }


def profile_from_dict(payload: object) -> FilterProfile:
    if not isinstance(payload, dict):
        raise ValueError("設定檔必須是 JSON object")
    payload = _migrate_profile_payload(payload)
    try:
        topics = tuple(
            ProfileTopic(
                name=str(topic["name"]),
                keywords=tuple(
                    KeywordDefinition(
                        phrase=str(keyword["phrase"]),
                        strength=KeywordStrength(keyword.get("strength", KeywordStrength.GENERAL.value)),
                    )
                    for keyword in topic["keywords"]
                ),
            )
            for topic in payload["topics"]
        )
        profile = FilterProfile(
            profile_id=str(payload["profile_id"]),
            name=str(payload["name"]),
            description=str(payload.get("description", "")),
            version=int(payload["version"]),
            selected_sources=tuple(str(source) for source in payload["selected_sources"]),
            topics=topics,
            minimum_score=int(payload.get("minimum_score", 3)),
        )
    except (KeyError, TypeError, ValueError) as exc:
        raise ValueError(f"設定檔格式錯誤：{exc}") from exc
    return validate_profile(profile)


def profiles_path() -> Path:
    if sys.platform == "darwin":
        root = Path.home() / "Library" / "Application Support"
    elif os.name == "nt":
        root = Path(os.environ.get("APPDATA", Path.home() / "AppData" / "Roaming"))
    else:
        root = Path(os.environ.get("XDG_CONFIG_HOME", Path.home() / ".config"))
    return root / "UKNewsScraper" / "profiles.json"


def load_profiles(path: str | Path | None = None) -> dict[str, FilterProfile]:
    destination = Path(path) if path else profiles_path()
    profiles = {DEFAULT_PROFILE_ID: default_profile()}
    if not destination.exists():
        return profiles
    try:
        payload = json.loads(destination.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError) as exc:
        raise ValueError(f"無法讀取設定檔：{exc}") from exc
    payload = _migrate_collection_payload(payload)
    for item in payload["profiles"]:
        profile = profile_from_dict(item)
        if profile.profile_id in profiles:
            raise ValueError(f"設定檔 ID 重複：{profile.profile_id}")
        profiles[profile.profile_id] = profile
    return profiles


def load_profiles_with_recovery(
    path: str | Path | None = None,
) -> ProfileLoadReport:
    destination = Path(path) if path else profiles_path()
    try:
        return ProfileLoadReport(load_profiles(destination))
    except ValueError as exc:
        recovery_path = _quarantine_invalid_file(destination)
        warning = f"{exc}；已載入內建設定"
        if recovery_path:
            warning += f"，原檔保留於 {recovery_path}"
        return ProfileLoadReport(
            {DEFAULT_PROFILE_ID: default_profile()},
            warning=warning,
            recovery_path=recovery_path,
        )


def save_profiles(profiles: list[FilterProfile], path: str | Path | None = None) -> Path:
    destination = Path(path) if path else profiles_path()
    validated = [validate_profile(profile) for profile in profiles if not profile.is_default]
    ids = [profile.profile_id for profile in validated]
    if len(ids) != len(set(ids)):
        raise ValueError("設定檔 ID 不可重複")
    payload = {
        "schema_version": PROFILE_SCHEMA_VERSION,
        "profiles": [profile_to_dict(profile) for profile in validated],
    }
    destination.parent.mkdir(parents=True, exist_ok=True)
    if destination.exists():
        backup = destination.with_suffix(destination.suffix + ".bak")
        backup_temporary = backup.with_suffix(backup.suffix + ".tmp")
        shutil.copy2(destination, backup_temporary)
        backup_temporary.replace(backup)
    temporary = destination.with_suffix(destination.suffix + ".tmp")
    temporary.write_text(json.dumps(payload, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")
    temporary.replace(destination)
    return destination


def restore_default_profiles(path: str | Path | None = None) -> dict[str, FilterProfile]:
    profile = default_profile()
    save_profiles([profile], path)
    return {DEFAULT_PROFILE_ID: profile}


def load_profile(reference: str | Path | None) -> FilterProfile:
    if reference is None:
        return default_profile()
    path = Path(reference).expanduser()
    if path.exists():
        return profile_from_dict(json.loads(path.read_text(encoding="utf-8")))
    profiles = load_profiles()
    try:
        return profiles[str(reference)]
    except KeyError as exc:
        raise ValueError(f"找不到設定檔：{reference}") from exc


def safe_profile_id(value: str) -> str:
    normalized = re.sub(r"[^a-z0-9-]+", "-", value.casefold()).strip("-")
    return normalized or "custom"


def _default_strength(keyword: str) -> KeywordStrength:
    normalized = keyword.casefold()
    if normalized in _SUPPORTING_KEYWORDS:
        return KeywordStrength.SUPPORTING
    if " " in keyword.strip() or len(keyword.strip()) >= 10:
        return KeywordStrength.CORE
    return KeywordStrength.GENERAL


def _migrate_collection_payload(payload: object) -> dict[str, Any]:
    if not isinstance(payload, dict) or not isinstance(payload.get("profiles"), list):
        raise ValueError("設定檔集合必須包含 profiles array")
    try:
        version = int(payload.get("schema_version", 0))
    except (TypeError, ValueError) as exc:
        raise ValueError("設定檔集合版本必須是整數") from exc
    if version > PROFILE_SCHEMA_VERSION:
        raise ValueError(f"不支援的設定檔集合版本：{version}")
    migrated = dict(payload)
    profiles = list(migrated["profiles"])
    if version == 0:
        profiles = [_migrate_profile_payload(item) for item in profiles]
        version = 1
    if version != PROFILE_SCHEMA_VERSION:
        raise ValueError(f"無法遷移設定檔集合版本：{version}")
    migrated["schema_version"] = version
    migrated["profiles"] = profiles
    return migrated


def _migrate_profile_payload(payload: dict[str, Any]) -> dict[str, Any]:
    migrated = dict(payload)
    try:
        version = int(migrated.get("version", 0))
    except (TypeError, ValueError) as exc:
        raise ValueError("設定檔版本必須是整數") from exc
    if version > PROFILE_SCHEMA_VERSION:
        raise ValueError(f"不支援的設定檔版本：{version}")
    if version == 0:
        migrated_topics = []
        for topic in migrated.get("topics", []):
            migrated_topic = dict(topic)
            migrated_topic["keywords"] = [
                (
                    {"phrase": keyword, "strength": KeywordStrength.GENERAL.value}
                    if isinstance(keyword, str)
                    else keyword
                )
                for keyword in migrated_topic.get("keywords", [])
            ]
            migrated_topics.append(migrated_topic)
        migrated["topics"] = migrated_topics
        migrated["version"] = 1
        version = 1
    if version != PROFILE_SCHEMA_VERSION:
        raise ValueError(f"無法遷移設定檔版本：{version}")
    selected_sources = migrated.get("selected_sources")
    if isinstance(selected_sources, list):
        expanded_sources: list[object] = []
        for source in selected_sources:
            replacements = SOURCE_ID_MIGRATIONS.get(str(source), (source,))
            for replacement in replacements:
                if replacement not in expanded_sources:
                    expanded_sources.append(replacement)
        migrated["selected_sources"] = expanded_sources
    return migrated


def _quarantine_invalid_file(path: Path) -> Path | None:
    if not path.exists():
        return None
    timestamp = datetime.now(timezone.utc).strftime("%Y%m%dT%H%M%SZ")
    recovery = path.with_name(f"{path.stem}.corrupt-{timestamp}{path.suffix}")
    try:
        path.replace(recovery)
    except OSError:
        return None
    return recovery

File: native/crates/uk-news-app/src/profiles.rs
use anyhow::{Context, Result};
use chrono::Utc;
use regex::Regex;
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use std::{
    collections::HashSet,
    fs,
    path::{Path, PathBuf},
};
use uk_news_core::{
    FilterProfile, KeywordDefinition, KeywordStrength, ProfileTopic, PROFILE_SCHEMA_VERSION,
};

const SUPPORTING: [&str; 7] = [
    "copyright",
    "evaluation",
    "innovation",
    "platform",
    "resilience",
    "supply chain",
    "technology",
];
const TOPICS: [(&str, &[&str]); 6] = [
    (
        "Science & Technology",
        &[
            "science",
            "scientific",
            "science and technology",
            "sci tech",
            "technology",
            "technological",
            "innovation",
            "research and development",
            "R&D",
        ],
    ),
    (
        "AI",
        &[
            "artificial intelligence",
            "AI",
            "machine learning",
            "foundation model",
            "foundation models",
            "frontier model",
            "frontier AI",
            "generative AI",
            "automated decision",
            "automated decision-making",
            "algorithm",
            "algorithmic accountability",
            "explainability",
            "fairness",
            "AI Safety Institute",
            "AI Security Institute",
            "model evaluation",
            "evaluation",
            "copyright",
            "text and data mining",
            "TDM",
            "privacy-by-design",
            "AI and data protection",
        ],
    ),
    (
        "資料治理/隱私/數位身份",
        &[
            "data protection",
            "UK GDPR",
            "GDPR",
            "Data Protection Act",
            "international data transfer",
            "adequacy",
            "standard contractual clauses",
            "data use",
            "data access",
            "open data",
            "data reuse",
            "data reuse policy",
            "Open Government Data Framework",
            "digital identity",
            "trust framework",
            "GOV.UK One Login",
            "One Login",
            "UK Digital Identity and Attributes Trust Framework",
        ],
    ),
    (
        "數位平台",
        &[
            "online safety",
            "Online Safety Act",
            "platform",
            "illegal content",
            "child safety",
            "recommendation algorithm",
            "algorithm transparency",
            "media plurality",
            "broadcast impartiality",
            "online influence transparency",
            "disinformation",
            "misinformation",
            "foreign information manipulation",
            "electoral information manipulation",
            "digital markets",
            "Digital Markets Competition and Consumers Act",
            "Strategic Market Status",
            "SMS",
            "systemic risk",
        ],
    ),
    (
        "網路安全/資安",
        &[
            "cyber",
            "cybersecurity",
            "network and information systems",
            "NIS",
            "critical infrastructure",
            "critical national infrastructure",
            "product security",
            "telecommunications infrastructure",
            "vulnerability",
            "coordinated vulnerability disclosure",
            "secure by design",
            "secure-by-design",
            "SBOM",
            "software bill of materials",
            "supply chain",
            "government cyber security strategy",
            "Cyber Essentials",
            "incident response",
            "resilience",
            "protective security",
            "hybrid threat",
            "hybrid threat resilience",
            "critical infrastructure protection",
            "national cyber resilience",
        ],
    ),
    (
        "半導體/量子技術",
        &[
            "semiconductor",
            "National Semiconductor Strategy",
            "semiconductor strategy",
            "chip",
            "microelectronics",
            "export control",
            "quantum",
            "National Quantum Strategy",
            "quantum technologies",
            "quantum computing",
            "quantum R&D",
        ],
    ),
];

pub fn default_profile() -> FilterProfile {
    FilterProfile {
        profile_id: "uk-tech-law".into(),
        name: "UK 科技法制".into(),
        description: "英國科技、AI、資料治理、平台、資安、半導體與量子政策觀測".into(),
        version: PROFILE_SCHEMA_VERSION,
        selected_sources: [
            "DSIT",
            "AISI",
            "ICO",
            "CMA",
            "UK IPO",
            "GDS",
            "Ofcom",
            "NCSC",
            "Electoral Commission",
            "Cabinet Office",
            "NPSA",
            "DBT",
            "UKRI",
            "UK Parliament",
        ]
        .into_iter()
        .map(Into::into)
        .collect(),
        topics: TOPICS
            .iter()
            .map(|(name, words)| ProfileTopic {
                name: (*name).into(),
                keywords: words
                    .iter()
                    .map(|word| KeywordDefinition {
                        phrase: (*word).into(),
                        strength: if SUPPORTING.iter().any(|x| x.eq_ignore_ascii_case(word)) {
                            KeywordStrength::Supporting
                        } else if word.trim().contains(' ') || word.trim().len() >= 10 {
                            KeywordStrength::Core
                        } else {
                            KeywordStrength::General
                        },
                    })
                    .collect(),
            })
            .collect(),
        minimum_score: 3,
    }
}

pub fn load_profile(value: Option<&str>) -> Result<FilterProfile> {
    let Some(value) = value else {
        return Ok(default_profile());
    };
    let path = Path::new(value);
    if path.exists() {
        let p: FilterProfile =
            serde_json::from_str(&fs::read_to_string(path)?).context("parse profile JSON")?;
        validate_profile(&p)?;
        Ok(p)
    } else {
        load_profiles()?
            .into_iter()
            .find(|p| p.profile_id == value)
            .with_context(|| format!("找不到設定檔：{value}"))
    }
}
pub fn validate_profile(p: &FilterProfile) -> Result<()> {
    if p.version != PROFILE_SCHEMA_VERSION {
        anyhow::bail!("不支援的設定檔版本：{}", p.version)
    }
    if !Regex::new(r"^[a-z0-9][a-z0-9-]{1,48}$")
        .unwrap()
        .is_match(&p.profile_id)
        || p.selected_sources.is_empty()
        || p.topics.is_empty()
        || p.minimum_score < 1
    {
        anyhow::bail!("設定檔欄位不完整")
    }
    let available = uk_news_sources::agencies()
        .into_iter()
        .map(|a| a.short_name)
        .chain(["UK Parliament".into()])
        .collect::<HashSet<_>>();
    if let Some(source) = p.selected_sources.iter().find(|x| !available.contains(*x)) {
        anyhow::bail!("設定檔包含未知來源：{source}")
    }
    let mut keywords = HashSet::new();
    for topic in &p.topics {
        if topic.name.trim().is_empty() || topic.keywords.is_empty() {
            anyhow::bail!("主題或關鍵詞不可空白")
        }
        for word in &topic.keywords {
            if word.phrase.trim().is_empty() || !keywords.insert(word.phrase.to_lowercase()) {
                anyhow::bail!("關鍵詞空白或重複：{}", word.phrase)
            }
        }
    }
    Ok(())
}

#[derive(Serialize, Deserialize)]
struct ProfileCollection {
    schema_version: u32,
    profiles: Vec<FilterProfile>,
}
#[derive(Debug, Serialize)]
pub struct ProfileLoadReport {
    pub profiles: Vec<FilterProfile>,
    pub warning: String,
    pub recovery_path: Option<PathBuf>,
}
pub fn profiles_path() -> PathBuf {
    #[cfg(target_os = "macos")]
    let root = home().join("Library/Application Support");
    #[cfg(target_os = "windows")]
    let root = std::env::var_os("APPDATA")
        .map(PathBuf::from)
        .unwrap_or_else(|| home().join("AppData/Roaming"));
    #[cfg(not(any(target_os = "macos", target_os = "windows")))]
    let root = std::env::var_os("XDG_CONFIG_HOME")
        .map(PathBuf::from)
        .unwrap_or_else(|| home().join(".config"));
    root.join("UKNewsScraper/profiles.json")
}
fn home() -> PathBuf {
    std::env::var_os("HOME")
        .map(PathBuf::from)
        .unwrap_or_else(|| PathBuf::from("."))
}
pub fn load_profiles() -> Result<Vec<FilterProfile>> {
    load_profiles_from(&profiles_path())
}
fn load_profiles_from(path: &Path) -> Result<Vec<FilterProfile>> {
    let mut profiles = vec![default_profile()];
    if !path.exists() {
        return Ok(profiles);
    }
    let raw: serde_json::Value =
        serde_json::from_str(&fs::read_to_string(path)?).context("無法讀取設定檔")?;
    let payload: ProfileCollection =
        serde_json::from_value(migrate_collection(raw)?).context("設定檔集合格式錯誤")?;
    if payload.schema_version > PROFILE_SCHEMA_VERSION {
        anyhow::bail!("不支援的設定檔集合版本：{}", payload.schema_version)
    }
    for p in payload.profiles {
        validate_profile(&p)?;
        if profiles.iter().any(|x| x.profile_id == p.profile_id) {
            anyhow::bail!("設定檔 ID 重複：{}", p.profile_id)
        }
        profiles.push(p)
    }
    Ok(profiles)
}
pub fn load_profiles_with_recovery() -> ProfileLoadReport {
    let path = profiles_path();
    match load_profiles_from(&path) {
        Ok(profiles) => ProfileLoadReport {
            profiles,
            warning: String::new(),
            recovery_path: None,
        },
        Err(error) => {
            let recovery_path = quarantine(&path).ok().flatten();
            let location = recovery_path
                .as_ref()
                .map(|value| format!("，原檔保留於 {}", value.display()))
                .unwrap_or_default();
            ProfileLoadReport {
                profiles: vec![default_profile()],
                warning: format!("{error:#}；已載入內建設定{location}"),
                recovery_path,
            }
        }
    }
}
fn migrate_collection(mut raw: serde_json::Value) -> Result<serde_json::Value> {
    let object = raw
        .as_object_mut()
        .context("設定檔集合必須是 JSON object")?;
    let version = object
        .get("schema_version")
        .and_then(serde_json::Value::as_u64)
        .unwrap_or(0);
    if version > PROFILE_SCHEMA_VERSION as u64 {
        anyhow::bail!("不支援的設定檔集合版本：{version}")
    }
    let profiles = object
        .get_mut("profiles")
        .and_then(serde_json::Value::as_array_mut)
        .context("設定檔集合必須包含 profiles array")?;
    if version == 0 {
        for profile in profiles {
            let profile = profile
                .as_object_mut()
                .context("設定檔必須是 JSON object")?;
            let profile_version = profile
                .get("version")
                .and_then(serde_json::Value::as_u64)
                .unwrap_or(0);
            if profile_version > PROFILE_SCHEMA_VERSION as u64 {
                anyhow::bail!("不支援的設定檔版本：{profile_version}")
            }
            if profile_version == 0 {
                if let Some(topics) = profile
                    .get_mut("topics")
                    .and_then(serde_json::Value::as_array_mut)
                {
                    for topic in topics {
                        if let Some(keywords) = topic
                            .get_mut("keywords")
                            .and_then(serde_json::Value::as_array_mut)
                        {
                            for keyword in keywords.iter_mut() {
                                if let Some(phrase) = keyword.as_str() {
                                    *keyword = serde_json::json!({
                                        "phrase": phrase,
                                        "strength": "general"
                                    });
                                }
                            }
                        }
                    }
                }
                profile.insert("version".into(), serde_json::json!(PROFILE_SCHEMA_VERSION));
                profile
                    .entry("description")
                    .or_insert_with(|| serde_json::json!(""));
                profile
                    .entry("minimum_score")
                    .or_insert_with(|| serde_json::json!(3));
            }
        }
        object.insert(
            "schema_version".into(),
            serde_json::json!(PROFILE_SCHEMA_VERSION),
        );
    }
    Ok(raw)
}
fn quarantine(path: &Path) -> Result<Option<PathBuf>> {
    if !path.exists() {
        return Ok(None);
    }
    let stamp = Utc::now().format("%Y%m%dT%H%M%SZ");
    let recovery = path.with_file_name(format!(
        "{}.corrupt-{stamp}.json",
        path.file_stem()
            .and_then(|value| value.to_str())
            .unwrap_or("profiles")
    ));
    fs::rename(path, &recovery)?;
    Ok(Some(recovery))
}
pub fn save_profile(profile: FilterProfile) -> Result<PathBuf> {
    validate_profile(&profile)?;
    if profile.profile_id == "uk-tech-law" {
        anyhow::bail!("內建設定檔不可覆寫")
    }
    let path = profiles_path();
    let mut profiles = load_profiles().unwrap_or_else(|_| vec![default_profile()]);
    profiles.retain(|p| p.profile_id != "uk-tech-law" && p.profile_id != profile.profile_id);
    profiles.push(profile);
    write_profiles(&path, &profiles)?;
    Ok(path)
}
pub fn delete_profile(id: &str) -> Result<bool> {
    if id == "uk-tech-law" {
        return Ok(false);
    }
    let path = profiles_path();
    let mut profiles = load_profiles()?;
    let before = profiles.len();
    profiles.retain(|p| p.profile_id == "uk-tech-law" || p.profile_id != id);
    if profiles.len() == before {
        return Ok(false);
    }
    write_profiles(
        &path,
        &profiles
            .into_iter()
            .filter(|p| p.profile_id != "uk-tech-law")
            .collect::<Vec<_>>(),
    )?;
    Ok(true)
}
fn write_profiles(path: &Path, profiles: &[FilterProfile]) -> Result<()> {
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent)?
    }
    if path.exists() {
        let backup = path.with_extension("json.bak");
        let tmp_backup = path.with_extension("json.bak.tmp");
        fs::copy(path, &tmp_backup)?;
        fs::rename(tmp_backup, backup)?
    }
    let tmp = path.with_extension("json.tmp");
    fs::write(
        &tmp,
        format!(
            "{}\n",
            serde_json::to_string_pretty(&ProfileCollection {
                schema_version: PROFILE_SCHEMA_VERSION,
                profiles: profiles.to_vec()
            })?
        ),
    )?;
    fs::rename(tmp, path)?;
    Ok(())
}
pub fn profile_hash(p: &FilterProfile) -> String {
    let value = serde_json::to_value(p).unwrap();
    let canonical = canonical_json(&value);
    format!("{:x}", Sha256::digest(canonical.as_bytes()))
}
fn canonical_json(v: &serde_json::Value) -> String {
    match v {
        serde_json::Value::Object(m) => {
            let mut keys = m.keys().collect::<Vec<_>>();
            keys.sort();
            format!(
                "{{{}}}",
                keys.into_iter()
                    .map(|k| format!(
                        "{}:{}",
                        serde_json::to_string(k).unwrap(),
                        canonical_json(&m[k])
                    ))
                    .collect::<Vec<_>>()
                    .join(",")
            )
        }
        serde_json::Value::Array(a) => format!(
            "[{}]",
            a.iter().map(canonical_json).collect::<Vec<_>>().join(",")
        ),
        _ => serde_json::to_string(v).unwrap(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn default_profile_hash_matches_python_contract() {
        assert_eq!(
            profile_hash(&default_profile()),
            "4155af3e893e9fdf2ccab4361e87548321ce62a2a7032de7ebb121c5bd72c915"
        );
    }

    #[test]
    fn migrates_python_v0_profile_collection() {
        let dir = tempfile::tempdir().unwrap();
        let path = dir.path().join("profiles.json");
        fs::write(&path, r#"{"profiles":[{"profile_id":"old-profile","name":"Old","selected_sources":["NCSC"],"topics":[{"name":"Cyber","keywords":["cyber"]}]}]}"#).unwrap();
        let profiles = load_profiles_from(&path).unwrap();
        assert_eq!(profiles[1].version, 1);
        assert_eq!(
            profiles[1].topics[0].keywords[0].strength,
            KeywordStrength::General
        );
    }
}
